### PR TITLE
SSR: Allow server rendering of paths with query args

### LINKED ISSUE #14159: Login: Enable SSR for Login page when `redirect_to` param provided
SSR works when users access https://calypso.live/log-in but not for https://calypso.live/log-in?redirect_to=https%3A%2F%2Fcalypso.live%2Fdevdocs%2Fwelcome.

It's a know limitation, we need to fix it. There are two places where the existing logic needs to get updated:

https://github.com/Automattic/wp-calypso/pull/13339/files#diff-cdf8be48e336c794267be91d81a3a610L114
https://github.com/Automattic/wp-calypso/pull/13339/files#diff-b2d6784517d1151d7e75f0287d0c6478R125

According to @seear this limitation was initially introduced to avoid caching issues with Themes pages. We can relax this for pages other than Themes section.

### PR DESCRIPTION
**Pull request by seear**
Fixes #14159

SSR and server caching of pages for any path with query args was previously disabled to avoid filling up the SSR caches with arbitrary theme search strings, for example:

`http://wordpress.com/themes?s=big`

The disadvantage to this is that pages with discrete query args such as login page with `redirect_to` param can never be server-rendered.

This change removes the no-query-arg condition for SSR, and adds a context field, `disableSSR`, so individual routes, such as /themes, can disable SSR for a condition such as query args.

**To Test**
* Be sure to rebuild after checking out, because the PR has server changes
* before building, use `export DEBUG="calypso:server-render"`
* Test various themes routes and check that paths with ? args do not server render (see build console)
* Test that themes routes without ? args server render as usual, for example http://calypso.localhost:3000/theme/mood
* Test login SSR routes

### CODE REVIEW

File: server/render/index.js [hunk 1]
```diff
@@ -111,8 +111,7 @@ export function serverRender( req, res ) {
 		// Send state to client
 		context.initialReduxState = pick( context.store.getState(), reduxSubtrees );
 		// And cache on the server, too
-		if ( isEmpty( context.query ) ) {
-			// Don't cache if we have query params
+		if ( context.disableSSR ) {
```
Comment (gziolo) [inline server/render/index.js]: I think it should be:

```js
if ( ! context.disableSSR ) {
```

:)

**Comment by gziolo** [comment]
This is what we exactly need for `Login` page and other similar use cases like Account Recovery, etc.

**Comment by stephanethomas** [comment]
I tested some login pages with query parameters (e.g. http://calypso.localhost:3000/log-in?redirect_to=/me) with [@gziolo's change](https://github.com/Automattic/wp-calypso/pull/16133#discussion_r126933921), and confirm they are now rendered server side.

File: server/render/index.js (hunk 1 above)
Comment (seear) [inline server/render/index.js]: 😅

**Comment by seear** [comment]
Thanks @gziolo, @stephanethomas!

I'd like to wait until @ockham gets a chance to look at this before merging.

**Comment by Tug** [comment]
I know of at least one issue which might arise with this change:
- The `/log-in/link/use` route needs to be removed from the isomorphic login section as it relies on query params. (/cc @jblz)

**Comment by jblz** [comment]
I think it makes more sense to to flip this change into more of a whitelist approach. That is to say, apply existing behavior & do not SSR when query args are present except where controllers declare it's ok.

There's a likelihood that something will be missed if we try to go through and flag all routes manually that need it disabled.

**Comment by seear** [comment]
> I think it makes more sense to to flip this change into more of a whitelist approach. That is to say, apply existing behavior & do not SSR when query args are present except where controllers declare it's ok.

Can I ask why it is not typically desirable to SSR login routes with query args?

In the themes case it is to avoid filling up the cache with arbitrary search strings, which are passed in the arg. I was thinking this might be unique to themes, but from your comment above, perhaps this is good general practice?

**Comment by jblz** [comment]
I'm definitely not an expert on SSR & could definitely be incorrect here, but the thought was that arbitrary query args could fill up the cache on any route even if the arg values aren't being used in the app.

**Comment by seear** [comment]
> arbitrary query args could fill up the cache on any route even if the arg values aren't being used in the app.

I had in mind that for log-ins, the query args might be some finite set of `redirect-to` locations, and it would be useful to server-render the page for all those cases? If you have a rough idea of how many combinations there should be we can simply increase the cache size by that many slots.

Reading through the thread again at p4TIVU-6Qj-p2, it seems others prefer to keep SSR off for qargs by default. I disagree, but I don't feel strongly about it, so I don't mind if we switch the logic in this PR. Doing that _would_ make for a safer PR.

**Comment by stephanethomas** [comment]
So thinking about it more, it's probably not a good idea to cache login pages with a `redirect_to` parameter. We have around between 7000 and 16000 login attempts per hour. While I don't know how many were performed with this parameter (and associated values) I suspect we might end up filling a non negligible part of the cache since the latter only stores [500 elements](https://github.com/Automattic/wp-calypso/blob/607b509defed85d1202f20fe64e8e1b5aa8fd026/server/state-cache/index.js#L8).

I wonder how bad it would be in term of performance to render a page server-side without caching. I like the `context.disableSSR` approach but maybe we just need something more granular that would allow us to specify when we want to cache a route (and by route I mean a path and some specific parameters and values).

**Comment by seear** [comment]
> We have around between 7000 and 16000 login attempts per hour 
...
> only stores 500 elements

Even if those  16000 hits are using, say, 300 slots, that would be a good use of the cache, right? We are only storing tiny amounts of state in that cache. That 500 limit is fairly arbitrary and can be increased appropriately.

> I wonder how bad it would be in term of performance to render a page server-side without caching.

My view is that if we are going to render server side, we may as well cache the results. We are not at all resource-constrained in terms of memory at present. Moreover, performing the initial server render _is_ resource intensive because it blocks the node process, meaning no other requests can be served during that time.

**Comment by stephanethomas** [comment]
Thanks for linking to this post @seear. I realize now that I was thinking of the approach suggested by @gziolo. The only data that I have regarding the possible set of parameters and values for the login page is pMz3w-6Sg-p2, which is not much. I could be wrong regarding my assumption. Still I wonder if we shouldn't decouple management of server-side rendering from caching.

**Comment by stephanethomas** [comment]
I see, that's a good point indeed.

**Comment by seear** [comment]
> Still I wonder if we shouldn't decouple management of server-side rendering from caching.

A quick bit of background—the server caches ([state cache](https://github.com/Automattic/wp-calypso/tree/607b509defed85d1202f20fe64e8e1b5aa8fd026/server/state-cache) and [render cache](https://github.com/Automattic/wp-calypso/blob/60142e74319922e40a9e69e8ef87157d61878d1c/server/render/index.js#L27)) exist only for the purpose of SSR.

**Comment by Tug** [comment]
Hum remember we're caching pages using the `pathname` here, thus query params are not included in the cache key.

**Comment by seear** [comment]
> Hum remember we're caching pages using the pathname here, thus query params are not included in the cache key.

Good point.

**Comment by seear** [comment]
> query params are not included in the cache key.

This can be changed if necessary with no ill-effect on themes, since themes routes will never SSR urls with query params.

**Comment by ockham** [comment]
Late to the party :slightly_smiling_face: 

> The disadvantage to this is that pages with discrete query args such as login page with `redirect_to` param can never be server-rendered.

This indicates that we might want to try a whitelist type of approach instead of a plain flag, where each section can specify known qargs to be cached? Empty or missing array means don't SSR at all.

**Comment by ockham** [comment]
Or maybe a predicate rather than an array to allow for more flexibility.

**Comment by pento** [comment]
I'm inclined to keep the cache key as the `pathname`, and whitelist query args that can be safely cached.

To use `redirect_to` as an example:

It looks like `WpcomLoginForm` is the only component that renders the content of of the redirect URL - everything else keeps it in the state, and doesn't make use of it until logging in, which will always be after all components are mounted, and ready to use.

In the case of `WpcomLoginForm`, it's only used once in the login flow, with a statically passed `redirectTo` prop.

So, for login, at least, the existence and value of `redirect_to` can be ignored when deciding whether to cache or not.

**Comment by gziolo** [comment]
> I'm inclined to keep the cache key as the pathname, and whitelist query args that can be safely cached.

Agreed on this one 👍 Cheers 😃

**Comment by Tug** [comment]
I agree we should keep using `pathname` for our cache keys. However, I'm not sure I agree with this:

> So, for login, at least, the existence and value of redirect_to can be ignored when deciding whether to cache or not.

Let's say we ignore `redirect_to` for `/log-in`, we will cache the response of the first request which comes in. The `redirect_to` param in that request could contain sensitive information such this user oauth2 flow params (client_id, email, ...). It will end up being cached and served for every following requests to every users and indexing bots.

I think we should define which routes are cacheable regardless of query params. In the case of `log-in` and `/log-in/link/use` we need to either make them cacheable (by moving sensitive data away from the page) or move them out of isomorphic sections (I'm not sure that can be done but we could also think about moving the isomorphic param at the route level)

**Comment by gziolo** [comment]
>  It will end up being cached and served for every following requests to every users and indexing bots.

We never expose `redirect_to` param in the HTML code, so it won't get cached. `redirect_to` is passed from Redux state using propos, but it is only consumed in the action handler that sends data to the server.

**Comment by Tug** [comment]
Right we don't, I think we use to at some point. I agree then, we should whitelist query args 1 by 1 and double check those otherwise we may expose sensitive data into the cache.

**Comment by stephanethomas** [comment]
I agree with Tug, sensitive data is a real concern. Even if today we don't render such information in our components it doesn't mean it will be always the case. Moreover I don't think blindly using the path name is bulletproof enough, since you could also use arbitrary values in [route parameters](https://expressjs.com/en/guide/routing.html#route-parameters) as well. So it looks to me it should be the controller's responsibility to determine if SSR should be enabled, and which cache key to use.

@seear, @ockham, how does it sound?

**Comment by seear** [comment]
> should be the controller's responsibility to determine if SSR should be enabled

This PR introduces that ability with the [flag](https://github.com/Automattic/wp-calypso/pull/16133/files#diff-cdf8be48e336c794267be91d81a3a610R79), but it seems from discussion above that a routing whitelist is preferred?

> and which cache key to use

Yeah, SSR worked this way at one point, with the controller setting the cache key in `context`. Currently, retrieval of cached state is one of the earliest things to happen, so it seems that would need to be postponed to allow the controller to set a custom key.

Maybe the best way forward from here is if #amber starts a fresh PR that meets the requirements of the login use case?

**Comment by stephanethomas** [comment]
> Maybe the best way forward from here is if #amber starts a fresh PR that meets the requirements of the login use case?

Makes sense indeed, let's do that :).

**Comment by ockham** [comment]
> Moreover I don't think blindly using the path name is bulletproof enough, since you could also use arbitrary values in [route parameters](https://expressjs.com/en/guide/routing.html#route-parameters) as well.

@stephanethomas I think (and was assuming when I wrote that code) that `req.path` includes route params ([reference](http://expressjs.com/de/4x/api.html#req.path)). (I don't think it could be well-defined it if it didn't.)

**Pull request closed without merging**